File: cli/internal/tui/pages/diffs.go
package pages

import (
	"fmt"
	"image/color"
	"strings"

	tea "charm.land/bubbletea/v2"
	"charm.land/lipgloss/v2"
	"github.com/niuulabs/volundr/cli/internal/api"
	tui "github.com/niuulabs/volundr/cli/internal/tui"
)

// DiffFilesLoadedMsg carries the list of changed files from the session pod.
type DiffFilesLoadedMsg struct {
	Files []api.DiffFileEntry
	Err   error
}

// DiffContentLoadedMsg carries the diff content for a single file.
type DiffContentLoadedMsg struct {
	Path string
	Diff string
	Err  error
}

// DiffFile represents a file in the diff tree.
type DiffFile struct {
	Path      string
	Status    string // "M" modified, "A" added, "D" deleted
	Diff      string
	Additions int
	Deletions int
}

// DiffsPage displays a file tree and diff viewer.
type DiffsPage struct {
	files     []DiffFile
	filtered  []int // indices into files that match the search
	cursor    int
	scrollPos int
	search    string
	searching bool
	loading   bool
	loadErr   error
	width     int
	height    int

	// Session pod client for fetching diffs.
	podClient *api.SessionPodClient
	session   *api.Session
	client    *api.Client
}

// NewDiffsPage creates a new diffs page.
func NewDiffsPage(client *api.Client) DiffsPage {
	return DiffsPage{
		client: client,
	}
}

// Init initializes the diffs page.
func (d DiffsPage) Init() tea.Cmd { //nolint:gocritic // value receiver needed for page interface consistency
	return nil
}

// SetSession configures the diffs page for a specific session and loads files.
func (d *DiffsPage) SetSession(sess api.Session) tea.Cmd { //nolint:gocritic // hugeParam acceptable for API type
	d.session = &sess
	d.files = nil
	d.cursor = 0
	d.scrollPos = 0
	d.loading = true
	d.loadErr = nil

	if sess.CodeEndpoint == "" {
		d.loading = false
		d.loadErr = fmt.Errorf("session has no code endpoint (status: %s)", sess.Status)
		return nil
	}

	token := ""
	if d.client != nil {
		token = d.client.Token()
	}
	d.podClient = api.NewSessionPodClient(sess.CodeEndpoint, token)
	client := d.podClient
	return func() tea.Msg {
		files, err := client.GetDiffFiles("last-commit")
		return DiffFilesLoadedMsg{Files: files, Err: err}
	}
}

// Update handles messages for the diffs page.
func (d DiffsPage) Update(msg tea.Msg) (DiffsPage, tea.Cmd) { //nolint:gocritic // value receiver needed for page interface consistency
	switch msg := msg.(type) {
	case DiffFilesLoadedMsg:
		d.loading = false
		if msg.Err != nil {
			d.loadErr = msg.Err
			return d, nil
		}
		d.files = nil
		for _, f := range msg.Files {
			d.files = append(d.files, DiffFile{
				Path:      f.Path,
				Status:    f.Status,
				Additions: f.Additions,
				Deletions: f.Deletions,
			})
		}
		d.applySearch()
		// Load diff content for the first file.
		if len(d.filteredFiles()) > 0 {
			return d, d.loadDiffContent(d.actualIndex(0))
		}
		return d, nil

	case DiffContentLoadedMsg:
		if msg.Err == nil {
			for i := range d.files {
				if d.files[i].Path == msg.Path {
					d.files[i].Diff = msg.Diff
					break
				}
			}
		}
		return d, nil

	case tea.KeyMsg:
		if d.searching {
			return d.handleSearchInput(msg)
		}

		filtered := d.filteredFiles()
		switch msg.String() {
		case "up", "k":
			if d.cursor > 0 {
				d.cursor--
				d.scrollPos = 0
				return d, d.loadDiffContent(d.actualIndex(d.cursor))
			}
		case "down", "j":
			if d.cursor < len(filtered)-1 {
				d.cursor++
				d.scrollPos = 0
				return d, d.loadDiffContent(d.actualIndex(d.cursor))
			}
		case "G":
			if len(filtered) > 0 {
				d.cursor = len(filtered) - 1
				d.scrollPos = 0
				return d, d.loadDiffContent(d.actualIndex(d.cursor))
			}
		case "g":
			if len(filtered) > 0 {
				d.cursor = 0
				d.scrollPos = 0
				return d, d.loadDiffContent(d.actualIndex(d.cursor))
			}
		case "J":
			d.scrollPos++
		case "K":
			if d.scrollPos > 0 {
				d.scrollPos--
			}
		case "/":
			d.searching = true
			d.search = ""
		case "r":
			if d.session != nil {
				cmd := d.SetSession(*d.session)
				return d, cmd
			}
		}
	}
	return d, nil
}

// loadDiffContent fetches the diff content for a file if not already loaded.
func (d DiffsPage) loadDiffContent(index int) tea.Cmd { //nolint:gocritic // value receiver needed for page interface consistency
	if index < 0 || index >= len(d.files) || d.podClient == nil {
		return nil
	}
	file := d.files[index]
	if file.Diff != "" {
		return nil // Already loaded.
	}
	client := d.podClient
	path := file.Path
	return func() tea.Msg {
		diff, err := client.GetFileDiff("last-commit", path)
		return DiffContentLoadedMsg{Path: path, Diff: diff, Err: err}
	}
}

// SetSize updates the page dimensions.
func (d *DiffsPage) SetSize(w, h int) {
	d.width = w
	d.height = h
}

// View renders the diffs page.
func (d DiffsPage) View() string { //nolint:gocritic // value receiver needed for page interface consistency
	theme := tui.DefaultTheme

	titleStyle := lipgloss.NewStyle().
		Foreground(theme.TextPrimary).
		Bold(true).
		MarginBottom(1)

	// No session
	if d.session == nil {
		return lipgloss.NewStyle().
			Width(d.width).
			Height(d.height).
			Padding(1, 2).
			Render(lipgloss.JoinVertical(lipgloss.Left,
				titleStyle.Render("◧ Diffs"),
				"",
				lipgloss.NewStyle().Foreground(theme.AccentAmber).Render("  Select a session first (press 1 to go to Sessions, then Enter)"),
			))
	}

	// Loading
	if d.loading {
		return lipgloss.NewStyle().
			Width(d.width).
			Height(d.height).
			Padding(1, 2).
			Render(lipgloss.JoinVertical(lipgloss.Left,
				titleStyle.Render("◧ Diffs"),
				"",
				lipgloss.NewStyle().Foreground(theme.AccentAmber).Render("  Loading diffs..."),
			))
	}

	// Error
	if d.loadErr != nil {
		return lipgloss.NewStyle().
			Width(d.width).
			Height(d.height).
			Padding(1, 2).
			Render(lipgloss.JoinVertical(lipgloss.Left,
				titleStyle.Render("◧ Diffs"),
				"",
				lipgloss.NewStyle().Foreground(theme.AccentRed).Render(fmt.Sprintf("  Error: %v  (r to retry)", d.loadErr)),
			))
	}

	// Search bar
	var searchBar string
	if d.searching {
		searchBar = lipgloss.NewStyle().
			Foreground(theme.AccentAmber).
			Render("/ " + d.search + "\u2588")
	} else if d.search != "" {
		searchBar = lipgloss.NewStyle().
			Foreground(theme.TextMuted).
			Render("Filter: " + d.search + "  (/ to edit)")
	}

	// Split: file tree (left) and diff viewer (right)
	// Reserve space: title(1) + margin(1) + stats(1) + blank(1) + content + blank(1) + hints(1) + padding(2)
	treeWidth := 35
	diffWidth := d.width - treeWidth - 8
	paneHeight := d.height - 9
	if searchBar != "" {
		paneHeight--
	}

	tree := d.renderFileTree(treeWidth, paneHeight)
	diff := d.renderDiffView(diffWidth, paneHeight)
	content := lipgloss.JoinHorizontal(lipgloss.Top, tree, "  ", diff)

	// Stats summary
	totalAdd := 0
	totalDel := 0
	for _, f := range d.files {
		totalAdd += f.Additions
		totalDel += f.Deletions
	}

	stats := fmt.Sprintf("  %s  %s  %s files changed",
		lipgloss.NewStyle().Foreground(theme.AccentEmerald).Render(fmt.Sprintf("+%d", totalAdd)),
		lipgloss.NewStyle().Foreground(theme.AccentRed).Render(fmt.Sprintf("-%d", totalDel)),
		lipgloss.NewStyle().Foreground(theme.TextMuted).Render(fmt.Sprintf("%d", len(d.files))),
	)

	// Navigation hints
	hintStyle := lipgloss.NewStyle().Foreground(theme.TextMuted)
	hints := hintStyle.Render("  j/k select  G/g bottom/top  J/K scroll diff  / search  r refresh")

	// Scroll position indicator
	filtered := d.filteredFiles()
	var scrollInfo string
	if d.cursor >= 0 && d.cursor < len(filtered) {
		file := filtered[d.cursor]
		if file.Diff != "" {
			totalLines := len(strings.Split(file.Diff, "\n"))
			scrollInfo = hintStyle.Render(fmt.Sprintf("  line %d/%d", d.scrollPos+1, totalLines))
		}
	}

	parts := []string{
		titleStyle.Render("◧ Diffs"),
		stats,
	}
	if searchBar != "" {
		parts = append(parts, searchBar)
	}
	parts = append(parts, "", content, "", hints+scrollInfo)

	return lipgloss.NewStyle().
		Width(d.width).
		Height(d.height).
		Padding(1, 2).
		Render(lipgloss.JoinVertical(lipgloss.Left, parts...))
}

// renderFileTree renders the file tree sidebar.
func (d DiffsPage) renderFileTree(width, height int) string { //nolint:gocritic // value receiver needed for page interface consistency
	theme := tui.DefaultTheme
	filtered := d.filteredFiles()

	if len(filtered) == 0 {
		return lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(theme.BorderSubtle).
			Width(width).
			Height(height).
			Padding(1, 1).
			Foreground(theme.TextMuted).
			Render("  No changes")
	}

	var items []string
	for i, f := range filtered {
		var statusColor color.Color
		switch f.Status {
		case "M":
			statusColor = theme.AccentAmber
		case "A":
			statusColor = theme.AccentEmerald
		case "D":
			statusColor = theme.AccentRed
		}

		statusStyle := lipgloss.NewStyle().Foreground(statusColor)
		pathStyle := lipgloss.NewStyle().Foreground(theme.TextSecondary)

		statsStr := lipgloss.NewStyle().Foreground(theme.TextMuted).
			Render(fmt.Sprintf("+%d -%d", f.Additions, f.Deletions))

		line := fmt.Sprintf(" %s %s %s",
			statusStyle.Render(f.Status),
			pathStyle.Render(truncatePath(f.Path, width-14)),
			statsStr,
		)

		if i == d.cursor {
			items = append(items, lipgloss.NewStyle().
				Background(theme.BgTertiary).
				Foreground(theme.AccentAmber).
				Width(width-2).
				Render(line))
		} else {
			items = append(items, lipgloss.NewStyle().
				Width(width-2).
				Render(line))
		}
	}

	content := strings.Join(items, "\n")

	return lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(theme.BorderSubtle).
		Width(width).
		Height(height).
		Render(content)
}

// renderDiffView renders the diff content for the selected file.
func (d DiffsPage) renderDiffView(width, height int) string { //nolint:gocritic // value receiver needed for page interface consistency
	theme := tui.DefaultTheme
	filtered := d.filteredFiles()

	if len(filtered) == 0 {
		return lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(theme.BorderSubtle).
			Width(width).
			Height(height).
			Render(lipgloss.NewStyle().Foreground(theme.TextMuted).Render("  No file selected"))
	}

	file := filtered[d.cursor]

	if file.Diff == "" {
		return lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(theme.BorderSubtle).
			Width(width).
			Height(height).
			Padding(1, 1).
			Render(lipgloss.NewStyle().Foreground(theme.TextMuted).Render("  Loading diff..."))
	}

	lines := strings.Split(file.Diff, "\n")

	// Apply scroll offset
	if d.scrollPos >= len(lines) {
		d.scrollPos = max(0, len(lines)-1)
	}

	visibleLines := height - 2
	start := d.scrollPos
	end := start + visibleLines
	if end > len(lines) {
		end = len(lines)
	}

	var coloredLines []string
	for _, line := range lines[start:end] {
		switch {
		case strings.HasPrefix(line, "+"):
			coloredLines = append(coloredLines,
				lipgloss.NewStyle().Foreground(theme.AccentEmerald).Render(line))
		case strings.HasPrefix(line, "-"):
			coloredLines = append(coloredLines,
				lipgloss.NewStyle().Foreground(theme.AccentRed).Render(line))
		case strings.HasPrefix(line, "@@"):
			coloredLines = append(coloredLines,
				lipgloss.NewStyle().Foreground(theme.AccentCyan).Render(line))
		default:
			coloredLines = append(coloredLines,
				lipgloss.NewStyle().Foreground(theme.TextSecondary).Render(line))
		}
	}

	content := strings.Join(coloredLines, "\n")

	return lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(theme.BorderSubtle).
		Width(width).
		Height(height).
		Padding(0, 1).
		Render(content)
}

// handleSearchInput processes keystrokes in search mode.
func (d DiffsPage) handleSearchInput(msg tea.KeyMsg) (DiffsPage, tea.Cmd) { //nolint:gocritic // value receiver needed for page interface consistency
	switch msg.String() {
	case "enter", "esc":
		d.searching = false
	case "backspace":
		if d.search != "" {
			d.search = d.search[:len(d.search)-1]
		}
	case "space":
		d.search += " "
	default:
		if text := msg.Key().Text; text != "" {
			d.search += text
		}
	}
	d.applySearch()
	return d, nil
}

// Searching returns whether the search input is active.
func (d DiffsPage) Searching() bool { //nolint:gocritic // value receiver needed for page interface consistency
	return d.searching
}

// applySearch rebuilds the filtered index list based on the search term.
func (d *DiffsPage) applySearch() {
	d.filtered = nil
	lower := strings.ToLower(d.search)
	for i, f := range d.files {
		if d.search != "" && !strings.Contains(strings.ToLower(f.Path), lower) {
			continue
		}
		d.filtered = append(d.filtered, i)
	}
	if d.cursor >= len(d.filtered) {
		d.cursor = max(0, len(d.filtered)-1)
	}
}

// filteredFiles returns the DiffFile entries matching the current search.
func (d DiffsPage) filteredFiles() []DiffFile { //nolint:gocritic // value receiver needed for page interface consistency
	if len(d.filtered) == 0 && d.search == "" {
		return d.files
	}
	result := make([]DiffFile, 0, len(d.filtered))
	for _, idx := range d.filtered {
		if idx < len(d.files) {
			result = append(result, d.files[idx])
		}
	}
	return result
}

// actualIndex maps a filtered cursor position to the real files index.
func (d DiffsPage) actualIndex(cursor int) int { //nolint:gocritic // value receiver needed for page interface consistency
	if d.search == "" || len(d.filtered) == 0 {
		return cursor
	}
	if cursor < 0 || cursor >= len(d.filtered) {
		return cursor
	}
	return d.filtered[cursor]
}

// truncatePath shortens a file path to fit within the given width.
func truncatePath(path string, maxLen int) string {
	if len(path) <= maxLen {
		return path
	}
	return "..." + path[len(path)-maxLen+3:]
}

File: src/volundr/tui/diffs.py
"""Diffs page — split view with file tree (left) and diff viewer (right)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from rich.markup import escape
from textual.app import ComposeResult
from textual.containers import Horizontal, Vertical
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Input, Static

from cli.tui.theme import (
    ACCENT_AMBER,
    ACCENT_CYAN,
    ACCENT_EMERALD,
    ACCENT_RED,
    TEXT_MUTED,
    TEXT_SECONDARY,
)


@dataclass
class DiffFile:
    """Represents a changed file with its diff content."""

    path: str = ""
    status: str = "M"  # M modified, A added, D deleted
    diff: str = ""
    additions: int = 0
    deletions: int = 0


def _status_color(status: str) -> str:
    match status:
        case "M":
            return ACCENT_AMBER
        case "A":
            return ACCENT_EMERALD
        case "D":
            return ACCENT_RED
        case _:
            return TEXT_MUTED


def _truncate_path(path: str, max_len: int) -> str:
    if len(path) <= max_len:
        return path
    return "..." + path[len(path) - max_len + 3 :]


def _colorize_diff_line(line: str) -> str:
    """Apply syntax coloring to a single diff line."""
    escaped = escape(line)
    if line.startswith("+"):
        return f"[{ACCENT_EMERALD}]{escaped}[/]"
    if line.startswith("-"):
        return f"[{ACCENT_RED}]{escaped}[/]"
    if line.startswith("@@"):
        return f"[{ACCENT_CYAN}]{escaped}[/]"
    return f"[{TEXT_SECONDARY}]{escaped}[/]"


class FileTreeItem(Widget):
    """A single file in the tree sidebar."""

    DEFAULT_CSS = """
    FileTreeItem { height: 1; padding: 0 1; }
    FileTreeItem.selected { background: #27272a; }
    """

    def __init__(self, file: DiffFile, *, selected: bool = False, **kwargs: Any) -> None:
        super().__init__(**kwargs)
        self._file = file
        if selected:
            self.add_class("selected")

    def compose(self) -> ComposeResult:
        f = self._file
        color = _status_color(f.status)
        stats = f"[{TEXT_MUTED}]+{f.additions} -{f.deletions}[/]"
        yield Static(
            f"[{color}]{f.status}[/] [{TEXT_SECONDARY}]{_truncate_path(f.path, 28)}[/] {stats}"
        )


class DiffsPage(Widget):
    """Split-pane diff viewer with file tree and syntax-colored diff.

    Keybindings:
        j/k     select file in tree
        J/K     scroll diff content
        G/g     jump to last/first file
        /       search files
        r       refresh
    """

    DEFAULT_CSS = """
    DiffsPage { width: 1fr; height: 1fr; }
    DiffsPage #diffs-header { height: auto; padding: 0 0 1 0; }
    DiffsPage #diffs-stats { height: 1; }
    DiffsPage #diffs-search { height: auto; display: none; }
    DiffsPage #diffs-search.visible { display: block; }
    DiffsPage #diffs-body { height: 1fr; }
    DiffsPage #diffs-tree {
        width: 36; border: round #27272a; overflow-y: auto;
    }
    DiffsPage #diffs-viewer {
        width: 1fr; border: round #27272a; overflow-y: auto; padding: 0 1;
    }
    DiffsPage #diffs-viewer-content { height: auto; }
    DiffsPage #diffs-empty { color: #71717a; padding: 2; }
    """

    cursor: reactive[int] = reactive(0)
    scroll_pos: reactive[int] = reactive(0)
    searching: reactive[bool] = reactive(False)

    def __init__(self, files: list[DiffFile] | None = None, **kwargs: Any) -> None:
        super().__init__(**kwargs)
        self._all_files: list[DiffFile] = files or []
        self._filtered: list[DiffFile] = list(self._all_files)
        self._search_term = ""
        self._mounted = False

    def compose(self) -> ComposeResult:
        with Vertical():
            yield Static("", id="diffs-stats")
            with Horizontal(id="diffs-search"):
                yield Input(placeholder="Filter files…", id="diffs-search-input")
            with Horizontal(id="diffs-body"):
                yield Vertical(id="diffs-tree")
                yield Vertical(id="diffs-viewer")

    def on_mount(self) -> None:
        self._mounted = True
        self._apply_filter()
        self._update_stats()

    # ── Data ────────────────────────────────────────────────

    def set_files(self, files: list[DiffFile]) -> None:
        """Replace all files and refresh."""
        self._all_files = list(files)
        self._search_term = ""
        self.cursor = 0
        self.scroll_pos = 0
        self._apply_filter()
        self._update_stats()

    # ── Filter ──────────────────────────────────────────────

    def _apply_filter(self) -> None:
        search = self._search_term.lower()
        if not search:
            self._filtered = list(self._all_files)
        else:
            self._filtered = [f for f in self._all_files if search in f.path.lower()]
        if self.cursor >= len(self._filtered):
            self.cursor = max(0, len(self._filtered) - 1)
        self._rebuild_tree()
        self._rebuild_viewer()

    def _update_stats(self) -> None:
        total_add = sum(f.additions for f in self._all_files)
        total_del = sum(f.deletions for f in self._all_files)
        text = (
            f"  [{ACCENT_EMERALD}]+{total_add}[/]  "
            f"[{ACCENT_RED}]-{total_del}[/]  "
            f"[{TEXT_MUTED}]{len(self._all_files)} files changed[/]"
        )
        try:
            self.query_one("#diffs-stats", Static).update(text)
        except Exception:
            pass

    # ── Tree ────────────────────────────────────────────────

    def _rebuild_tree(self) -> None:
        if not self._mounted:
            return
        try:
            tree = self.query_one("#diffs-tree", Vertical)
        except Exception:
            return
        tree.remove_children()
        if not self._filtered:
            tree.mount(Static(f"[{TEXT_MUTED}]  No changes[/]"))
            return
        for i, f in enumerate(self._filtered):
            tree.mount(FileTreeItem(f, selected=(i == self.cursor)))

    # ── Diff viewer ─────────────────────────────────────────

    def _rebuild_viewer(self) -> None:
        if not self._mounted:
            return
        try:
            viewer = self.query_one("#diffs-viewer", Vertical)
        except Exception:
            return
        viewer.remove_children()
        if not self._filtered or self.cursor >= len(self._filtered):
            viewer.mount(Static(f"[{TEXT_MUTED}]  No file selected[/]"))
            return
        f = self._filtered[self.cursor]
        if not f.diff:
            viewer.mount(Static(f"[{TEXT_MUTED}]  Loading diff…[/]"))
            return
        lines = f.diff.split("\n")
        start = self.scroll_pos
        end = min(len(lines), start + 50)
        colored = "\n".join(_colorize_diff_line(ln) for ln in lines[start:end])
        viewer.mount(Static(colored))

    # ── Cursor / scroll ──────────────────────────────────────

    def watch_cursor(self) -> None:
        self.scroll_pos = 0
        self._rebuild_tree()
        self._rebuild_viewer()

    def watch_scroll_pos(self) -> None:
        self._rebuild_viewer()

    def action_cursor_up(self) -> None:
        if self.cursor > 0:
            self.cursor -= 1

    def action_cursor_down(self) -> None:
        if self.cursor < len(self._filtered) - 1:
            self.cursor += 1

    def action_cursor_top(self) -> None:
        self.cursor = 0

    def action_cursor_bottom(self) -> None:
        self.cursor = max(0, len(self._filtered) - 1)

    def action_scroll_diff_down(self) -> None:
        self.scroll_pos += 1

    def action_scroll_diff_up(self) -> None:
        if self.scroll_pos > 0:
            self.scroll_pos -= 1

    # ── Search ──────────────────────────────────────────────

    def action_toggle_search(self) -> None:
        self.searching = not self.searching

    def watch_searching(self, value: bool) -> None:
        try:
            box = self.query_one("#diffs-search", Horizontal)
        except Exception:
            return
        if value:
            box.add_class("visible")
            try:
                self.query_one("#diffs-search-input", Input).focus()
            except Exception:
                pass
        else:
            box.remove_class("visible")

    def on_input_changed(self, event: Input.Changed) -> None:
        if event.input.id == "diffs-search-input":
            self._search_term = event.value
            self._apply_filter()

    def action_refresh(self) -> None:
        self._apply_filter()
        self._update_stats()
